### PR TITLE
Move to go-spiffe/v2 API

### PR DESCRIPTION
**Pull request by faisal-memon**
This PR updates spiffe-helper pkg to use go-spiffe/v2 API. It adds in the ability to pass in a logger that is then passed to the workloadapi to better catch errors when bad sockets are defined.

### CODE REVIEW

**Comment by blaggacao** [comment]
@azdagron Is this considerable for merging? I think we'd want to be using the helper in all sorts of situations that don't yet support the workload-api natively.

File: cmd/spiffe-helper/main.go
Comment (azdagron) [inline cmd/spiffe-helper/main.go:42]: nit: context should in general be cancelled by defer (thought it doesn't really matter here since this is `main()`)
```suggestion
	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	defer cancel()
```

File: pkg/sidecar/sidecar.go
Comment (azdagron) [inline pkg/sidecar/sidecar.go:214]: This change pre-allocates a number of bytes equal to the number of certs being written, which doesn't make sense.

File: pkg/sidecar/sidecar.go
Comment (azdagron) [inline pkg/sidecar/sidecar.go:187]: x509.MarshalPKCS8PrivateKey takes `interface {}` so this type assertion isn't necessary.

File: pkg/sidecar/sidecar.go
Comment (azdagron) [inline pkg/sidecar/sidecar.go:188]: Though unlikely to fail, we should check for and report an error from here defensively.

File: pkg/sidecar/sidecar.go
Comment (azdagron) [inline pkg/sidecar/sidecar.go:254]:
```suggestion
		w.sidecar.config.Log.Errorf("Watching x509 context: %v", err)
```

File: go.mod
Comment (azdagron) [inline go.mod:9]: Hmm, where are we still relying on v1? If possible we should get rid of this dependency.

File: go.mod
Comment (azdagron) [inline go.mod:10]: Due to the _very_ long time to review this, we've moved quite far from this beta.2 release. Can we grab the latest?

**Review by azdagron — COMMENTED** [review]
Sorry for the _very late_ review on this, @faisal-memon. I'd still be happy to take it if you have cycles to address the comments. If not, I can probably take a little time to get it to the finish line.

**Comment by azdagron** [comment]
Yes, @blaggacao, we can certainly take this. Sorry it has taken so long to get around to review. The spiffe-helper doesn't really have official maintainers and is often forgotten.

**Comment by blaggacao** [comment]
@azdagron Thank you, in the nearer future, I might be adding JWT-SVID facilities. But haven't come around to make up my mind about a design.

We need to "patch" a lot of legacy components out of the hashicorp stack to work with SPIFFR so we'll be depending on the helper for some time, still.

**Comment by faisal-memon** [comment]
> Sorry for the _very late_ review on this, @faisal-memon. I'd still be happy to take it if you have cycles to address the comments. If not, I can probably take a little time to get it to the finish line.

I can address them.

File: go.mod
Comment (faisal-memon) [reply go.mod:9]: We are using `github.com/spiffe/go-spiffe/spiffetest` which is pulling in this dependancy.

**Comment by blaggacao** [comment]
@azdagron would the latest modifications be merge-worthy?

**Comment by azdagron** [comment]
Maybe? Though there has been some movement on go-spiffe (the official v2 version was released) that we'd need to pull and fix up (either in this PR or a follow-up).

Unfortunately, the SPIFFE organization has tightened up permissions and this component is currently without maintainers and I no longer have access. I've raised this with the @spiffe/ssc.

**Comment by joshuastern** [comment]
Any update on this? This is something my organization is interested in using.

**Comment by azdagron** [comment]
We're still happy to take this change under consideration. It needs to be pulled up to the latest go-spiffe v2 version (this was pre-release). @faisal-memon are you still willing to take this changes under your wing? I can follow up with the @spiffe/ssc about repository rights. Also, I wonder if it is worth just pulling in and forking the spiretest package from the v1 version of go-spiffe into this codebase to get rid of that dependency as well.

**Comment by faisal-memon** [comment]
> We're still happy to take this change under consideration. It needs to be pulled up to the latest go-spiffe v2 version (this was pre-release). @faisal-memon are you still willing to take this changes under your wing? I can follow up with the @spiffe/ssc about repository rights. Also, I wonder if it is worth just pulling in and forking the spiretest package from the v1 version of go-spiffe into this codebase to get rid of that dependency as well.

Yea, ill bring it up the latest version.

File: go.mod
Comment (faisal-memon) [reply go.mod]: dependency removed

File: .go-version
Comment (azdagron) [inline .go-version:1]: If we're going to bump this, can we move to the latest patch version? I don't know if anybody is using this on Windows, but 1.19.3 contains a security fix for environment variable handling in spawned processes on Windows.

File: cmd/spiffe-helper/main.go
Comment (azdagron) [inline cmd/spiffe-helper/main.go:42]: Why are we applying a timeout to the context used to start the daemon? This will cause the watcher run in the goroutine of RunDaemon to stop watching after five seconds. Also, since we select on this context below, the process will exit.

File: cmd/spiffe-helper/main.go
Comment (azdagron) [inline cmd/spiffe-helper/main.go:52]: Instead of exporting this channel, why can't we make RunDaemon blocking, move the error channel here, kick off the goroutine here in main, and simply send down the error channel with the result of RunDaemon? Exposing an exported error channel here feels a little clunky. One of the things patterns we try to avoid  is the "start/stop" pattern (this was a motivator to rewrite go-spiffe v1), instead preferring blocking functions and leaving it up to callers to decide if they need to spawn a goroutine to call it. Sometimes they do, sometimes they don't. Start/stop forces callers into a specific concurrency pattern.

File: cmd/spiffe-helper/main.go
Comment (azdagron) [inline cmd/spiffe-helper/main.go:50]: We should instead use signal.NotifyContext to cancel the context when the signal is hit.

File: pkg/sidecar/sidecar.go
Comment (azdagron) [inline pkg/sidecar/sidecar.go:74]: nit: If we're just using this client once, to start up a watcher, it would be better to use the top-level [WatchX509Context](https://pkg.go.dev/github.com/spiffe/go-spiffe/v2@v2.1.1/workloadapi#WatchX509Context)  convenience function.

File: pkg/sidecar/sidecar.go
Comment (azdagron) [inline pkg/sidecar/sidecar.go:246]: This log line is misleading. The x509 context could have been updated for myriad reasons (e.g., bundle changes, single SVID change).

File: pkg/sidecar/sidecar.go
Comment (azdagron) [inline pkg/sidecar/sidecar.go:255]:
```suggestion
		w.sidecar.config.Log.Errorf("Error while watching x509 context: %v", err)
```

File: cmd/spiffe-helper/main.go
Comment (faisal-memon) [reply cmd/spiffe-helper/main.go:42]: The timeout functionality is existing. I think this was an attempt to move this functionality over from the v1 client. Here is the relevant snippet from the current code base: https://github.com/spiffe/spiffe-helper/blob/3c14013b4adf3b622f95b9367c4842ab0d26b806/pkg/sidecar/sidecar.go#L157-L160

I cant find `X509ClientConfig`, was it moved or deprecated?

Comment (azdagron) [reply cmd/spiffe-helper/main.go:42]: The workload package referenced there isn't a go-spiffe v1 thing, but rather a different workload api client package that existed in SPIRE. The timeout there would cause a failure to be returned from the Start() method if a successful response had not been received from the workload api in that time. The existing sidecar code just logs errors returned from start and tries again after some backoff. We don't need to replicate that in the new code since we are already logging errors in the watcher implementation and the workload api client internally performs reasonable backoff.

**Comment by azdagron** [comment]
Considering the lack of integration tests for this woefully unmaintained component, as part of authoring this change, would you mind giving this a pretty thorough manual test of the impacted code? I feel like the timeout bug I pointed out that shut down the helper in five seconds would have been caught through that kind of testing.

File: pkg/sidecar/sidecar.go
Comment (azdagron) [inline pkg/sidecar/sidecar.go:63]: Changing this to an unbuffered channel breaks the current semantics provided by `CertReadyChan`. The only callers in this repository are tests, which reach in here to replace this channel (which is a bit gross), but who knows if other folks are using this package. Is there a reason for this change that I'm not seeing?

**Review by azdagron — COMMENTED** [review]
Looks pretty good, faisal. I have two remaining comments. Also, looks like we're missing DCO signoff on a bunch of the commits?

File: pkg/sidecar/sidecar.go
Comment (faisal-memon) [reply pkg/sidecar/sidecar.go]: I don't remember for sure, but i think its not necessary to have the buffer so thats why i removed it. I put it back because we should probably have a notice for that type of change.

File: cmd/spiffe-helper/main.go
Comment (faisal-memon) [reply cmd/spiffe-helper/main.go]: done

File: pkg/sidecar/sidecar.go
Comment (azdagron) [reply pkg/sidecar/sidecar.go]: If it isn't buffered then a race is introduced between callers of CertReadyChan and updateCertificates. If consumers of  CertReadyChan are not selecting on the channel before updateCertificates tries to insert into certReadyChan then the insert will be dropped (because of the `default`) and consumers won't be notified until the next cert update (even though they should have been notified for the original update). It for sure needs to be buffered as currently designed.

Comment (faisal-memon) [reply pkg/sidecar/sidecar.go]: yes, thats a good point.

**Comment by azdagron** [comment]
Still looking for fixups on DCO.

**Comment by faisal-memon** [comment]
> Still looking for fixups on DCO.

Done. Is there a way to modify the "suggested change" feature to include a DCO? Thats where alot of them get missed.

**Review by azdagron — APPROVED** [review]

**Pull request merged**
